### PR TITLE
IndexOutOfRange Exception in KeyToVector transformer

### LINKED ISSUE #2678: Unintuitive error when giving array to OneHotEncoder
The following will throw a very ugly exception only when calling `Preview()` and not before:

```csharp
    public class MlDataEx
    {
      public string[] A { get; set; }
      public MlDataEx(string a)
      {
        this.A = new[] { a };
      }
    }

    public static void MlStuff()
    {
      var ctx = new MLContext();
      var dv = ctx.Data.ReadFromEnumerable(new[] { new MlDataEx("A") });
      var data = ctx.Transforms.Categorical.OneHotEncoding("A").Fit(dv).Transform(dv).Preview();
    }
```

### PR DESCRIPTION
**Pull request by Ivanidzo4ka**
Fixes #2678

### CODE REVIEW

File: src/Microsoft.ML.Data/Transforms/KeyToVector.cs [hunk 1]
```diff
@@ -573,7 +573,6 @@ private ValueGetter<VBuffer<float>> MakeGetterInd(DataViewRow input, int iinfo)
                                 if (key >= (uint)size)
                                     continue;
                                 editor.Values[count] = 1;
-                                editor.Indices[count++] = slot * size + (int)key;
```
Comment (eerhardt) [inline src/Microsoft.ML.Data/Transforms/KeyToVector.cs]: test?

Comment (eerhardt) [reply src/Microsoft.ML.Data/Transforms/KeyToVector.cs]: It also appears that there are tests failing in CI with this change.

Maybe the right change is to use 

https://github.com/dotnet/machinelearning/blob/01a362b9a4a31a774f4f59bbe942d78240a69c7d/src/Microsoft.ML.Core/Data/VBufferEditor.cs#L42-L44

**Comment by TomFinley** [comment]
Hi @Ivanidzo4ka, is this relevant to #2185? It seems to be touching much of the same code I noted in that issue.

I mentioned you there but perhaps you did not see.

File: src/Microsoft.ML.Data/Transforms/Hashing.cs [hunk 2]
```diff
@@ -1273,7 +1273,7 @@ public SchemaShape GetOutputSchema(SchemaShape inputSchema)
                     metadata.Add(slotMeta);
                 if (colInfo.InvertHash != 0)
                     metadata.Add(new SchemaShape.Column(AnnotationUtils.Kinds.KeyValues, SchemaShape.Column.VectorKind.Vector, TextDataViewType.Instance, false));
-                result[colInfo.Name] = new SchemaShape.Column(colInfo.Name, col.ItemType is VectorType ? SchemaShape.Column.VectorKind.Vector : SchemaShape.Column.VectorKind.Scalar, NumberDataViewType.UInt32, true, new SchemaShape(metadata));
+                result[colInfo.Name] = new SchemaShape.Column(colInfo.Name, col.Kind, NumberDataViewType.UInt32, true, new SchemaShape(metadata));
             }
```
Comment (wschin) [inline src/Microsoft.ML.Data/Transforms/Hashing.cs]: Does it mean input shape must be scalar? If yes, we need to throw when encountering a vector.

**Comment by wschin** [comment]
>                     throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", colInfo.inputColumnName);

We don't we just check if the input is a key?

---
Refers to: src/Microsoft.ML.Transforms/KeyToVectorMapping.cs:461 in 540b139. [](commit_id = 540b139656bc5c5d6ce90b5055e1549f4fb38f60, deletion_comment = False)

File: src/Microsoft.ML.Transforms/KeyToVectorMapping.cs [hunk 3]
```diff
@@ -466,7 +466,9 @@ public override SchemaShape GetOutputSchema(SchemaShape inputSchema)
                         metadata.Add(new SchemaShape.Column(AnnotationUtils.Kinds.SlotNames, SchemaShape.Column.VectorKind.Vector, keyMeta.ItemType, false));
                 if (col.Kind == SchemaShape.Column.VectorKind.Scalar)
                     metadata.Add(new SchemaShape.Column(AnnotationUtils.Kinds.IsNormalized, SchemaShape.Column.VectorKind.Scalar, BooleanDataViewType.Instance, false));
-                result[colInfo.outputColumnName] = new SchemaShape.Column(colInfo.outputColumnName, SchemaShape.Column.VectorKind.Vector, NumberDataViewType.Single, false, new SchemaShape(metadata));
+                result[colInfo.outputColumnName] = new SchemaShape.Column(colInfo.outputColumnName,
+                    col.Kind == SchemaShape.Column.VectorKind.VariableVector ? SchemaShape.Column.VectorKind.VariableVector : SchemaShape.Column.VectorKind.Vector,
```
Comment (wschin) [inline src/Microsoft.ML.Transforms/KeyToVectorMapping.cs]: >col [](start = 20, length = 3)

This is key-to-vector mapping. Can the input column be a vector-valued?

File: src/Microsoft.ML.Data/Transforms/Hashing.cs (hunk 2 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Data/Transforms/Hashing.cs]: No. It means we accept vectors, scalars and varvectors

---
In reply to: [263166153](https://github.com/dotnet/machinelearning/pull/2681#discussion_r263166153) [](ancestors = 263166153)

File: src/Microsoft.ML.Transforms/KeyToVectorMapping.cs (hunk 3 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Transforms/KeyToVectorMapping.cs]: You can have vector of keys, right?

---
In reply to: [263169008](https://github.com/dotnet/machinelearning/pull/2681#discussion_r263169008) [](ancestors = 263169008)

Comment (TomFinley) [reply src/Microsoft.ML.Transforms/KeyToVectorMapping.cs]: In fact a vector of keys is one of the most important scenarios (e.g., bag of words).

File: src/Microsoft.ML.Data/Transforms/Hashing.cs [hunk 4]
```diff
@@ -1273,7 +1273,7 @@ public SchemaShape GetOutputSchema(SchemaShape inputSchema)
                     metadata.Add(slotMeta);
                 if (colInfo.InvertHash != 0)
                     metadata.Add(new SchemaShape.Column(AnnotationUtils.Kinds.KeyValues, SchemaShape.Column.VectorKind.Vector, TextDataViewType.Instance, false));
-                result[colInfo.Name] = new SchemaShape.Column(colInfo.Name, col.ItemType is VectorType ? SchemaShape.Column.VectorKind.Vector : SchemaShape.Column.VectorKind.Scalar, NumberDataViewType.UInt32, true, new SchemaShape(metadata));
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/Transforms/Hashing.cs]: >col.ItemType is VectorType ? SchemaShape.Column.VectorKind.Vector : SchemaShape.Column.VectorKind [](start = 76, length = 97)

Geez, the original author sure got confused.

File: src/Microsoft.ML.Data/Transforms/KeyToVector.cs [hunk 5]
```diff
@@ -793,14 +793,16 @@ public override SchemaShape GetOutputSchema(SchemaShape inputSchema)
...
                         metadata.Add(new SchemaShape.Column(AnnotationUtils.Kinds.SlotNames, SchemaShape.Column.VectorKind.Vector, keyMeta.ItemType, false));
                 if (!colInfo.Bag && (col.Kind == SchemaShape.Column.VectorKind.Scalar || col.Kind == SchemaShape.Column.VectorKind.Vector))
                     metadata.Add(new SchemaShape.Column(AnnotationUtils.Kinds.CategoricalSlotRanges, SchemaShape.Column.VectorKind.Vector, NumberDataViewType.Int32, false));
                 if (!colInfo.Bag || (col.Kind == SchemaShape.Column.VectorKind.Scalar))
                     metadata.Add(new SchemaShape.Column(AnnotationUtils.Kinds.IsNormalized, SchemaShape.Column.VectorKind.Scalar, BooleanDataViewType.Instance, false));
 
-                result[colInfo.Name] = new SchemaShape.Column(colInfo.Name, SchemaShape.Column.VectorKind.Vector, NumberDataViewType.Single, false, new SchemaShape(metadata));
+                result[colInfo.Name] = new SchemaShape.Column(colInfo.Name,
+                    col.Kind == SchemaShape.Column.VectorKind.VariableVector && !colInfo.Bag ? SchemaShape.Column.VectorKind.VariableVector : SchemaShape.Column.VectorKind.Vector,
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/Transforms/KeyToVector.cs]: >col.Kind == SchemaShape.Column.VectorKind.VariableVector && !colInfo.Bag ? SchemaShape.Column.VectorKind.VariableVector : SchemaShape.Column.VectorKind.Vector [](start = 20, length = 158)

Um... yeah.if you wanted to explode out this condition a bit with a bunch of clarifying comments I wouldn't object to this becoming more than a one-line statement. I found this *slightly* hard to unpack.

**Comment by TomFinley** [comment]
> Hi @Ivanidzo4ka, is this relevant to #2185? It seems to be touching much of the same code I noted in that issue.

Hi, @Ivanidzo4ka, wondering if you've had a chance to give this any thought or attention.

**Comment by eerhardt** [comment]
@Ivanidzo4ka - what's the status of this PR? Can it be updated?

**Comment by Ivanidzo4ka** [comment]
@eerhardt  I've update it.

File: test/BaselineOutput/SingleRelease/Categorical/featurized.tsv [hunk 6]
```diff
@@ -1,14 +0,0 @@
-#@ TextLoader{
```
Comment (eerhardt) [inline test/BaselineOutput/SingleRelease/Categorical/featurized.tsv]: Can the `SingleDebug` versions of these two baseline files be deleted as well?

File: test/Microsoft.ML.Tests/Transformers/CategoricalTests.cs [hunk 7]
```diff
@@ -297,12 +283,15 @@ public void TestCommandLine()
...
+            var data = new[] {
+                new TestClass() { A = 1, B = new int[2] { 2, 3 }, C = new int[2] { 3, 4 } },
+                new TestClass() { A = 4, B = new int[2] { 2, 4 }, C = new int[3] { 2, 4, 3 } }
+            };
             var dataView = ML.Data.LoadFromEnumerable(data);
             var pipe = ML.Transforms.Categorical.OneHotEncoding(new[]{
-                    new InputOutputColumnPair("TermA", "A"),
-                    new InputOutputColumnPair("TermB", "B"),
-                    new InputOutputColumnPair("TermC", "C")
+                    new OneHotEncodingEstimator.ColumnOptions("CatA", "A"),
```
Comment (eerhardt) [inline test/Microsoft.ML.Tests/Transformers/CategoricalTests.cs]: `OneHotEncodingEstimator.ColumnOptions` is an internal class.  Can we instead write the tests using the public API?

**Comment by codemzs** [comment]
closing this PR since #3763 has these changes in it. I had initially created that PR not knowing there was this PR but then @yaeldekel pointed out this PR also takes care of output schema so I took those changes. I need to check-in my PR as it is blocking 1P customer.

**Pull request closed without merging**